libraries/AP_Telemetry/AP_Telemetry_MQTT.cpp

```diff
@@ -0,0 +1,252 @@
+/*
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "AP_Telemetry_MQTT.h"
+#include "define_MQTT.h"
+#include <stdio.h>
+
+#include <stdlib.h>
+#include <sys/types.h>
+#include <pthread.h>
+#include <time.h>
+#include <string>
+
+#include <getopt.h>
+
+extern const AP_HAL::HAL& hal;
+
+extern int mqtt_to_mavlink_message(const char* cmd, mavlink_message_t *msg);
+
+MQTTAsync AP_Telemetry_MQTT::mqtt_client;
+AP_Telemetry_MQTT* AP_Telemetry_MQTT::telemetry_mqtt = nullptr;
+MQTTAsync_connectOptions AP_Telemetry_MQTT::conn_options = MQTTAsync_connectOptions_initializer;
+char const* AP_Telemetry_MQTT::mqtt_server;
+int mqtt_msg_arrived(void *context, char *topicname, int topicLen, MQTTAsync_message* message);
+void onConnect(void *context, MQTTAsync_successData* response);
+void onConnectFailure(void* context, MQTTAsync_failureData* response);
+
+// MQTT Client accessor
+MQTTAsync* AP_Telemetry_MQTT::get_MQTTClient()
+{
+    return &mqtt_client;
+}
+
+AP_Telemetry_MQTT* AP_Telemetry_MQTT::get_telemetry_mqtt()
+{
+    return telemetry_mqtt;
+}
+
+AP_Telemetry_MQTT* AP_Telemetry_MQTT::init_telemetry_mqtt(AP_Telemetry &frontend, AP_HAL::UARTDriver* uart)
+{
+    if (telemetry_mqtt == nullptr) {
+        telemetry_mqtt = new AP_Telemetry_MQTT(frontend, uart);
+        telemetry_mqtt->init_mqtt();
+    }
+    return telemetry_mqtt;
+}
+
+AP_Telemetry_MQTT::AP_Telemetry_MQTT(AP_Telemetry &frontend, AP_HAL::UARTDriver* uart) :
+    AP_Telemetry_Backend(frontend, uart)
+{}
+
+void AP_Telemetry_MQTT::init_mqtt()
+{
+    int rc;
+    mqtt_mutex_store = PTHREAD_MUTEX_INITIALIZER;
+    mqtt_mutex = &mqtt_mutex_store;
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+
+    recv_msg_list = ListInitialize();
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if ((rc = pthread_mutex_init(mqtt_mutex, &attr)) != 0) {
+        printf("init: error %d initializing mqtt_mutex\n", rc);
+        MQTTHandle_error(rc);
+    } else {
+
+        conn_options.keepAliveInterval = 20;
+        conn_options.cleansession = 1;
+
+        conn_options.onSuccess = onConnect;
+        conn_options.onFailure = onConnectFailure;
+        conn_options.context = mqtt_client;
+
+        char clientid[9] = "no_id";
+        srand((unsigned int)time(0));
+        sprintf(clientid, "client_%d", rand()%100);
```

Comment:
The client id is used to connect to connect to the mqtt server and should be unique.
This part should be upgrade, but I don't have idea about how to do. The client have to decide an id before connect to the server(message broker) but only the message broker could assert the chosen id is free or not.